EuiSuperSelect

cchaos opened the pull request:
Addresses #274 

----

This creates a custom `EuiPopover` implementation that styles the triggering button to look like a `EuiSelect` and populates a list of `EuiContextMenuItem`s based on an `options` object prop.

<img width="442" alt="screen shot 2018-06-11 at 20 00 50 pm" src="https://user-images.githubusercontent.com/549577/41314645-60e8e9e6-6e5b-11e8-977a-993bd7cfa2cb.png">

The issue right now is that it still behaves like a popover and not a select. Things that need help:

- [ ] Utilize the same keyboard functionality as in `EuiComboBox`
- [ ] Fix the focus/active states of the triggering button
- [ ] _Probably_ better positioning either from `EuiComboBox` or whatever the fix is going to be for `EuiPopover`
- [ ] _Question_: Does it need the `isInvalid` stuff as the typical `EuiSelect` does? If so, what does it do and how will it need to be implemented?

cchaos commented:
@chandlerprall Do you think you could help out with the JS stuff for this one?

snide commented:
> Does it need the isInvalid stuff as the typical EuiSelect does? If so, what does it do and how will it need to be implemented

Yes, we'll need it. Though I think it should just come from `EuiFormRow` the same as any other. You'll just want to pass some form of `isInvalid` prop into this component as well. I'm guessing that will do literally nothing other than make the underline red.

chandlerprall commented:
@cchaos yeah, no problem! I can get to it later today / tomorrow

cchaos commented:
@chandlerprall, Any chance you will be able to get back to this this week?

chandlerprall commented:
@cchaos yes! Sorry, fell off my radar.

cchaos commented:
@chandlerprall Sorry, I merged your width PR in without testing. I noticed that when the popover is closing it doesn't maintain the determined width: https://d.pr/free/v/xzqbKp

cchaos commented:
@chandlerprall thoughts on the above?

cchaos commented:
Ok, this is finally no longer a WIP. Ready for review!

snide commented on src/components/form/super_select/super_select.js:
```diff
@@ -0,0 +1,250 @@
+      e.preventDefault();
+      e.stopPropagation();
+      this.shiftFocus(SHIFT_BACK);
+    } else if (e.keyCode === keyCodes.DOWN) {
```
This works great, love that it goes back to the top. I'd also add a key event for "space" that does the same as down so that it mirrors the behavior of a HTML select.

chandlerprall replied:
Does `SPACE` not work as expected for you? For me, if the select box is focused but not open `SPACE` opens the options list, if an option is focused `SPACE` selects that option and closes the dropdown.

cchaos replied:
`SPACE` works for me

snide replied:
Yeah, sorry, chatted this in slack. Was a bad report on my side. This works as expected.

snide commented on src/components/popover/_popover.scss:
```diff
@@ -131,6 +131,11 @@
   }
+
+
+  &.euiPopover__panel-noArrow .euiPopover__panel__arrow {
```
Element of an element? Maybe just make this something unique?

cchaos replied:
I don't understand.

snide replied:
Sorry, should have been more direct.

`.euiPopover__panel__arrow`

I think we've tried to avoid using underscores twice in a single selector. My read of this is that arrow is an element of the element panel of the block popover.

Should it just be....

`euiPopover__panelArrow` as the element of popover instead?

cchaos replied:
Yeah, that was leftover from another PR, but I can change it here

snide commented on src/components/context_menu/_context_menu_item.scss:
```diff
@@ -34,3 +34,19 @@
     align-self: flex-end;
   }
+
+  .euiContextMenu__itemLayout {
```
Note, that I would spend 5 minutes on this. I don't think it's a big deal, was just something small i noticed.

cchaos replied:
The whole popover is fuzzy (even the text). I don't think it's related to the drop shadow stuff. I'll look into it though.

chandlerprall replied:
Could very well be from fractional numbers on popover positioning. LMK if that's the issue.

cchaos replied:
Hmm, my local looks fine:

<img width="357" alt="screen shot 2018-08-06 at 13 43 52 pm" src="https://user-images.githubusercontent.com/549577/43732167-d5836396-997e-11e8-853f-38d2221a24f8.png">

cchaos replied:
Even with decimal placement it is fine:

<img width="726" alt="screen shot 2018-08-06 at 13 46 14 pm" src="https://user-images.githubusercontent.com/549577/43732287-3b57de0e-997f-11e8-853d-3ff7c2e67f0e.png">


Though, @chandlerprall maybe the positioning service should just go ahead and round to a whole number?

cjcenizal commented:
Wow, the globalToastList tests failed... now THAT'S weird.

```
19:12:24 FAIL src/components/toast/global_toast_list.test.js
19:12:24   ● EuiGlobalToastList › props › dismissToast › is called when the toast lifetime elapses
19:12:24 
19:12:24     expect(received).toBe(expected)
19:12:24     
19:12:24     Expected value to be (using Object.is):
19:12:24       true
19:12:24     Received:
19:12:24       false
19:12:24 
19:12:24        96 |         // The callback is invoked once the toast fades from view.
19:12:24        97 |         setTimeout(() => {
19:12:24     >  98 |           expect(dismissToastSpy.called).toBe(true);
19:12:24        99 |           done();
19:12:24       100 |         }, TOAST_LIFE_TIME_MS + TOAST_FADE_OUT_MS + 10);
19:12:24       101 |       });
19:12:24       
19:12:24       at src/components/toast/global_toast_list.test.js:98:42
19:12:24       at Timeout.callback [as _onTimeout] (node_modules/jsdom/lib/jsdom/browser/Window.js:633:19)
19:12:24 
19:12:24   ● EuiGlobalToastList › props › dismissToast › is called when the toast lifetime elapses
19:12:24 
19:12:24     Error: Uncaught [Error: expect(received).toBe(expected)
19:12:24     
19:12:24     Expected value to be (using Object.is):
19:12:24       true
19:12:24     Received:
19:12:24       false]
19:12:24 
19:12:24       2 | // failed prop types check.
19:12:24       3 | console.error = message => {
19:12:24     > 4 |   throw new Error(message);
19:12:24       5 | };
19:12:24       6 | 
19:12:24       
19:12:24       at reportException (node_modules/jsdom/lib/jsdom/living/helpers/runtime-script-errors.js:66:24)
19:12:24       at Timeout.callback [as _onTimeout] (node_modules/jsdom/lib/jsdom/browser/Window.js:635:7)
19:12:24       at BufferedConsole.Object.<anonymous>.console.error (scripts/jest/setup/throw_on_console_error.js:4:9)
19:12:24       at VirtualConsole.on.e (node_modules/jsdom/lib/jsdom/virtual-console.js:29:45)
19:12:24       at reportException (node_modules/jsdom/lib/jsdom/living/helpers/runtime-script-errors.js:70:28)
19:12:24       at Timeout.callback [as _onTimeout] (node_modules/jsdom/lib/jsdom/browser/Window.js:635:7)
```

cjcenizal commented:
Retest

cjcenizal commented:
Jenkins test this?

cjcenizal commented on src/components/form/super_select/_super_select_control.scss:
```diff
@@ -0,0 +1,36 @@
+ * 2. Ensure the descenders don't get cut off
+ */
+
+ .euiSuperSelectControl {
```
I think there's an extra space at the beginning of this line.

cchaos replied:
Yeah, my IDE doesn't like those comment blocks so anything afterward it tries to align to the comment.

cjcenizal commented on src/components/form/super_select/_super_select.scss:
```diff
@@ -0,0 +1,34 @@
+  }
+}
+
+.euiSuperSelect__popoverPanel[class*="bottom"] {
```
What does this do? If it's anticipating a particular class, then maybe it makes more sense to use that specific class name(s) here so it's easier to find the reference if you change that class name?

cchaos replied:
So instead of relying on a full classname that is created by a different component like `.euiPopover__panel-bottom`, all I care about is the direction portion `bottom`. So if the prefix to the class name ever changes, it won't affect this component that depends on it.

cjcenizal replied:
I think it's important to remember we're also writing this codebase for everybody (Kibana team at large and community contributors)... so we should think about the point of view coming into this code and trying to figure out what it does so that they can either debug something or submit a PR.  If someone comes to the super select and reads this line, they need to reverse engineer how the super select works in conjunction with the popover to understand what classes it could be interacting with. This line gives the reader a clue, and then the reader needs to spend time figuring it out.

But if it was just `.euiSuperSelect__popoverPanel.euiPopover__panel-bottom` then the line tells the reader exactly what it does, and they can focus their time on their PR rather than trying to decipher the code.

cchaos replied:
I don't think your concern is about needing to "reverse engineer" than just now knowing what that selector means. This type of selector is a very common one and we shouldn't need to be explicit (and therefore too dependent) for those that just aren't as CSS knowledgable. Its a quick lookup and you can easily see the result or debug in the browser's dev tools.

cjcenizal replied:
I don't think this is worth blocking on, so I'm OK with merging this PR without this change. But if you want to know more about my motivations, here's a great article on the importance of [writing code so it can be understood](https://blog.codinghorror.com/when-understanding-means-rewriting/) (not saying that you don't care about this already, but I just thought it's worth sharing).

cchaos replied:
What I'll do is just add a comment about where that's coming from so then at least it's understandable.

cjcenizal commented on src-docs/src/views/super_select/super_select_complex.js:
```diff
@@ -0,0 +1,83 @@
+      {
+        value: 'option_one',
+        inputDisplay: 'Option one',
+        dropdownDisplay: (
```
Instead of putting `inputDisplay` and `dropdownDisplay` on each option, I think it might make more sense to treat these options as data-only, without any responsibility for rendering. Then we could provide props to the `EuiSuperSelect`  instance itself to render the input and the dropdown. This will allow these options to be defined by an API response without requiring us to then decorate each option. This is similar to how `EuiComboBox` allows you customize the way each option is rendered (https://github.com/elastic/eui/blob/master/src-docs/src/views/combo_box/render_option.js#L117).

I'm thinking along these lines:

```jsx
const healthToColorMap = {
  0: 'warning',
  1: 'subdued',
  2: 'danger',
};

const healthToNameMap = {
  0: 'Warning',
  1: 'Minor',
  2: 'Critical',
};

export default class extends Component {
  constructor(props) {
    super(props);

    this.options = [
      {
        value: 'option_one',
        // You can put anything you want inside of data.
        data: {
          name: 'Option one',
          description: 'Has a short description giving more detail to the option.',
          health: 0,
        },
      },
      {
        value: 'option_two',
        data: {
          name: 'Option two',
          description: 'Has a short description giving more detail to the option.',
          health: 1,
        },
      },
      {
        value: 'option_three',
        data: {
          name: 'Option three',
          description: 'Has a short description giving more detail to the option.',
          health: 2,
        },
      },
    ];

    this.state = {
      value: this.options[1].value,
    };
  }

  onChange = (value) => {
    this.setState({
      value: value,
    });
  };

  renderOption = option => {
    const { name, description, health } = option.data;

    return (
      <Fragment>
        <strong>{name}</strong>

        <EuiHealth color={healthToColorMap[health]} style={{ lineHeight: 'inherit' }}>
          healthToNameMap[health]
        </EuiHealth>

        <EuiSpacer size="xs" />
        <EuiText size="s" color="subdued">
          <p className="euiTextColor--subdued">{description}</p>
        </EuiText>
      </Fragment>
    );
  };

  renderSelectedOption = option => {
    return option.data.name;
  };

  render() {
    return (
      <Fragment>
        <EuiSuperSelect
          options={this.options}
          valueOfSelected={this.state.value}
          onChange={this.onChange}
          itemLayoutAlign="top"
          renderOption={this.renderOption}
          renderSelectedOption={this.renderSelectedOption}
          hasDividers
          aria-label="Use aria labels when no actual label is in use"
        />
      </Fragment>
    );
  }
}
```

chandlerprall replied:
I think it's valuable to have render-purposeful _show this in the select box_ and _show this in the drop down_ values. However, I would agree that also allowing a `data` param to help track the represented data would be helpful.

cchaos replied:
My worry with this option, is that each option is no longer customizable, unless you add a bunch of `if` statements in the render methods. You could still do a similar thing with the current mapping by using the render methods as calls within the options declaration.

snide replied:
Example would be two options in a dropdown. One preferred, one not. Think a legacy option that you want to de-emphasize, not use an icon for, color dirrectly...etc. I can think of a bunch of times we'd do something like that. Either way, as @cchaos mentioend, you could just do the render separately if you were looking for something more uniform. Generally we seem to benefit from flexibility in display, and always add it in later, which is why I like this format.

cchaos replied:
Like @snide said, it's more about having the flexibility to have different renders than restricting them all to needing to render the same or have all the same layout elements. Here's an instance:
![image](https://user-images.githubusercontent.com/549577/43726823-2bc71d6c-996e-11e8-8b18-0354af5154a2.png)

Where you might need custom styles applied to each item because of the difference in how each typeface may render.

cjcenizal replied:
I don't want to hold up this PR so I'm throwing in the towel on this point. 😄 But I just want to point out that in your example @cchaos, you'd just specify a `fontName` property on the `data` object and then map that to a className or a style to apply the correct font to that item. Each item is still essentially the same.

chandlerprall replied:
@cjcenizal and I discussed this and I'm mostly onboard with his suggestion. There's something nagging me that I'm not able to pinpoint, so I'll think about this some more.

Want to reassure @snide and @cchaos that there is no inherent loss of flexibility or customization with this approach. There's a few ways (CJ mentioned one) that retains this.

chandlerprall replied:
In the end, I do think the original implementation matches the intent of this component the closest. We want to be able to embrace differences in the options, and while this is possible with normalized `renderOption` methods, I think it is best served with the desired element contents passed directly.

cjcenizal replied:
@bmcconaghy and I found a possible use case for super select in the ILM feature, but it would require supporting hundreds of options. Will this component support that use caase? Will we be able to implement infinite scrolling and search? Think of the GitHub dropdown for finding users or labels:

![image](https://user-images.githubusercontent.com/1238659/44106879-a7a339a6-9faa-11e8-838b-8973b301836f.png)

cchaos replied:
As it is right now, no. But it can be altered down the road to support this.

cjcenizal replied:
Thanks @cchaos though I was more thinking from the technical perspective, e.g. is there a challenge with decorating hundreds (or thousands) of options with `inputDisplay` and `dropdownDisplay` properties, would it play well with `react-virtualized`, etc?

snide replied:
@cjcenizal That screen looks like EuiFilterGroup, which has a current implementation within our table / search components (which you can see on those pages respectively). If even adds search once it gets beyond 10 items. Will that not cover what you need? I don't know about it's scaling abilities, but it's a 1-1 design against the screen you showed, so my guess is that's the component we'd update further.

By design EuiSuperSelect should not have search in it and is really just a stylized select of choices. We have combo box and the filter group stuff for the pattern you describe.

cjcenizal replied:
@snide Good call! Yes that could work, but it needs to be broken out into its own component. Looks like the table example you shared has it baked into its interface via the `search` prop, and the standalone [EuiFilterGroup](https://elastic.github.io/eui/#/forms/filter-group) example is just using a hand-rolled popover.

If we were to break it out into its own component, then at first glance it looks like the only difference between it and EuiSuperSelect would be that one has search and the other doesn't.

cjcenizal commented on src/components/form/super_select/super_select_control.test.js:
```diff
@@ -0,0 +1,89 @@
+        <EuiSuperSelectControl
+          options={[
+            { value: '1', inputDisplay: 'Option #1', disabled: false },
+            { value: '2', inputDisplay: 'Option #2', disabled: true }
```
Can we add tests for passing `className` and `data-test-subj` to the options?

cchaos commented on src-docs/src/views/super_select/super_select_example.js:
```diff
@@ -0,0 +1,95 @@
+    text: (
+      <div>
+        <p>
+          This is a simple replacement component for <EuiCode>EuiSelect</EuiCode> if you
```
Added.

chandlerprall commented:
> I did run into a problem with screen-reader accessibility. When I tab to the super select, VoiceOver on OS X doesn’t read the selected option. When I tab to a regular select element, the selected option is read aloud first, and then the element itself is identified. Can we fix this so the behaviors are identical/similar?

@cchaos any thoughts here? It seems to me that to get similar Mac VoiceOver results we'd want to integrate an actual select box (which opens another issue, the `select` element that is included for form submission has `display: none;` which excludes its value from inclusion in form submissions). We probably could do something around building dynamic aria labels around the whole menu content, but the screen reader will still add ", button", etc, at the end.

snide commented:
I think we can probably solve it with some onFocus announcements. Specifically i don't think it matters if it's an actual select vs. a collection of buttons, and a role="group" with a proceeding announcement should work fine.

cchaos commented:
I'll defer to whatever choice makes it work.

cchaos commented:
@chandlerprall All there is now, for addressing feedback, revolves around the discussion of needing to provide `inputDisplay` **and** `dropdownDisplay`.

cjcenizal commented on src-docs/src/views/super_select/super_select_complex.js:
```diff
@@ -0,0 +1,82 @@
+
+    this.state = {
+      value: this.options[1].value,
+      inputDisplay: this.options[1].inputDisplay,
```
Looks like this isn't being consumed anywhere, so I think we can remove it?

cjcenizal commented on src-docs/src/views/super_select/super_select_complex.js:
```diff
@@ -0,0 +1,82 @@
+  onChange = (value) => {
+    this.setState({
+      value: value,
+      inputDisplay: this.options.inputDisplay,
```
Then we can simplify this to:

```js
this.setState({ value });
```

cjcenizal commented on src/components/form/super_select/super_select.test.js:
```diff
@@ -0,0 +1,82 @@
+      .toMatchSnapshot();
+  });
+
+  describe('props', () => {
```
I think we need to mount the components in these tests and then click to open them or manually set the open state. We can then use the `takeMountedSnapshot` utility to get a snapshot of it. Otherwise the snapshots will just consist of closed super selects. Here's an example of a similar test with the combo box: https://github.com/elastic/eui/blob/master/src/components/combo_box/combo_box.test.js#L75

cjcenizal commented on src/components/form/super_select/super_select_control.test.js:
```diff
@@ -0,0 +1,93 @@
+  });
+
+  describe('props', () => {
+    test('options are rendered', () => {
```
Doesn't look like this test is effective since the options list isn't open in the snapshot. But I also think it might not be necessary, since you're testing it indirectly in the super select test suite.
